lib/cpp/ipc/hid.cpp

```diff
@@ -0,0 +1,41 @@
+#include<libtransistor/cpp/types.hpp>
+#include<libtransistor/cpp/ipc/hid.hpp>
+#include<libtransistor/types.h>
+#include<libtransistor/ipc/hid.h>
+
+#include<neither/neither.hpp>
+
+namespace Transistor {
+namespace IPC {
+
+neither::Either<ResultCode, HID> HID::Initialize() {
+	HID obj;
+	return ResultCode::Maybe(hid_ipc_init()).rightMap([obj](auto const &v) -> HID { return obj; } );
+}
+
+HID::HID(const HID& other) {
+	// add another reference
+	ResultCode::AssertOk(hid_ipc_init());
+}
+
+HID::HID(HID&& other) {
+	// add another reference
+	ResultCode::AssertOk(hid_ipc_init());
+}
+
+HID::HID() { // should only be called from HID::Initialize()
+}
+
+HID::~HID() {
+	hid_ipc_finalize();
+}
+
+neither::Either<ResultCode, std::shared_ptr<KSharedMemory>> HID::GetSharedMemory() {
+	shared_memory_h handle;
+	return ResultCode::Maybe(hid_ipc_get_shared_memory_handle(&handle)).rightMap([&handle](auto const &v) -> std::shared_ptr<KSharedMemory> {
+			return std::make_shared<KSharedMemory>(handle);
+		});
+}
```

Comment:
I don't like using `shared_ptr` here, since I think that using such a thing is a decision that the caller should be able to make, but I also couldn't find a way to get the `KSharedMemory` object out without it or a copy of it getting destructed and closing the handle, which raises the question of how a `KObject` should behave when it is copied. Should it increase some reference count for the object? At that point, you may as well use `std::shared_ptr`, which is the solution I chose. Can we explicitly disable copying and force it to always be moved? I have a lot to learn about such memory management in modern C++. I keep trying to apply the concepts I picked up from Rust to C++ but don't know how.

Reply:
I think it is possible to disable copy constructors if you want it to behave like a singleton that can only be moved.